src/Sparrow/Platform/Posix/Syscall.cs

```diff
@@ -333,16 +340,88 @@ public static bool CheckSyncDirectoryAllowed(string filepath)
-                return -1;
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to open directory " + dir + " in order to sync for file " + file);
+            }
             var fsyncRc = FSync(fd);
             if (fsyncRc == -1)
-                return -1;
-            return close(fd);
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to FSync directory " + dir + " with fd=" + fd + " in order to sync for file " + file);
+            }
+
+            if (close(fd) == -1)
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to close directory " + dir + " with fd=" + fd + " in order to sync for file " + file);
+            }
+
+            if (isRealPath)
+                return 0;
+
+            // now sync the real path if the above is a symlink
+            var buffSize = 64;
+            var realpathToDir = ArrayPool<byte>.Shared.Rent(buffSize);
+            try
+            {
+                // determine buffSize passed to readlink should be using lstat (but this is not available right now)
+                // so we try with rented array of 64 bytes and if readlink needs more we will try again with 8K and fail otherwise
+                // (TODO: change it to lstat / FileInfo(path).Target in the future when https://github.com/dotnet/corefx/issues/26310 is fixed)
+                int numOfBytes;
+                fixed (byte* pBuff = realpathToDir)
+                {
+                    numOfBytes = ReadLinkOrThrow(dir, pBuff, buffSize);
+                    if (numOfBytes == 0)
+                        return 0; // non symlink dir
+ 
+                    if (numOfBytes >= buffSize) // 64 bytes are not enough for path legnth. lets try with 8192
+                    {
+                        buffSize = 8192; // usually PATH_MAX is 4096
+                        ArrayPool<byte>.Shared.Return(realpathToDir);
+                        realpathToDir = ArrayPool<byte>.Shared.Rent(buffSize);
+                        
+                        fixed (byte* pBuff8K = realpathToDir)
+                        {
+                            numOfBytes = ReadLinkOrThrow(dir, pBuff8K, buffSize);
+                            if (numOfBytes == 0)
+                                return 0; // non symlink dir
+                            
+                            if (numOfBytes >= buffSize)
+                            {
+                                throw new InvalidOperationException("readlink to " + dir +
+                                                                    " failed due to the need of more then 8192 bytes for path legnth. Failed to sync directory");
+                            }
+                        }
+                    }
+                }
+
+                var realpathString = Encoding.UTF8.GetString(realpathToDir,0, numOfBytes - 1);
```

Comment:
why -1 here?

Reply:
0 to Length -1 .. zero based array.

Reply:
numBytes came from `readlink`, which write the number of bytes that were written